### PR TITLE
Added support for PHPUnit 6.1

### PR DESCRIPTION
**Pull request by HeahDude**
My final blocker for migrating an app from Symfony 2.6 / PHPUnit 4.6 / PHP 5.4, to Symfony 2.8 / PHPUnit 6.1 / PHP 7.1.

Is this ok for you? Thanks!

### CODE REVIEW

**Comment by alexislefebvre** [comment]
Is there a solution to keep compatibility with PHPUnit 4.8?

IIRC there is a compatibility layer for `\PHPUnit_Framework_*` in PHPUnit 5, for forward compatibility with PHPUnit 6. But this doesn't exist in PHPUnit 4.8.

It looks like we'll have to drop support of PHPUnit 4.8 if we want to support PHPUnit6.

**Comment by HeahDude** [comment]
Looks like it's fixed now :).

**Comment by alexislefebvre** [comment]
Yes they backported the forward compatibility to PHPUnit 4.8: https://github.com/sebastianbergmann/phpunit/blob/master/ChangeLog-4.8.md#added

File: Test/ValidationErrorsConstraint.php [hunk 1]
```diff
@@ -11,9 +11,13 @@
 
 namespace Liip\FunctionalTestBundle\Test;
 
+// BC
+class_alias('\PHPUnit_Framework_Constraint', '\PHPUnit\Framework\Constraint\Constraint');
```
Comment (alexislefebvre) [inline Test/ValidationErrorsConstraint.php]: @HeahDude This may be a stupid question but is it still necessary after your last commit that require PHPUnit `^4.8.35|…`?

Comment (HeahDude) [inline Test/ValidationErrorsConstraint.php]: Yes this requirement is for the low deps, this is what makes tests passing ;).

Comment (HeahDude) [inline Test/ValidationErrorsConstraint.php]: Are, got ya, saw the wrong diff, I guess you're right, let's test that :).

File: Test/ValidationErrorsConstraint.php [hunk 2]
```diff
@@ -11,9 +11,6 @@
 
 namespace Liip\FunctionalTestBundle\Test;
 
-// BC
-class_alias('\PHPUnit_Framework_Constraint', '\PHPUnit\Framework\Constraint\Constraint');
```
Comment (alexislefebvre) [inline Test/ValidationErrorsConstraint.php]: Okay, my previous comment was a bad idea, we need that, even if I don't know why. :)

**Comment by HeahDude** [comment]
Ping :)

**Comment by alexislefebvre** [comment]
I don't understand why none of the Travis CI builds installed PHPUnit 7.1. Do you know why Composer installed only PHPUnit 4.8 or 5.7?

**Comment by alexislefebvre** [comment]
I tested your branch on my local environment, PHPUnit 7.1 installed succesfully but tests of this Bundle failed.

I had to perform some changes so that tests passed, here is the full diff: https://gist.github.com/alexislefebvre/c439db16d52e0cc7a116417e7f7ccec9

Strangely, it looks like I have an error with `class_alias('\PHPUnit_Framework_Constraint', '\PHPUnit\Framework\Constraint\Constraint');`, probably because the aliased class does not exist in PHPUnit 6.1.

**Comment by alexislefebvre** [comment]
It looked like PHPUnit 6.1 wasn't installed because we used an old version of Ubuntu on Travis CI. I updated to Trusty and now I have some errors with MySQL :disappointed: : https://github.com/alexislefebvre/LiipFunctionalTestBundle/pull/49

**Comment by vpassapera** [comment]
I use this bundle with phpunit 6.1 with no problems.

Here is my composer file:

```
{
    "name": "allegiant/isengard",
    "license": "proprietary",
    "type": "project",
    "description": "Devops as a service.",
    "minimum-stability": "dev",
    "prefer-stable": true,
    "autoload": {
        "psr-4": {
            "": "src/"
        },
        "classmap": [
            "app/AppKernel.php",
            "app/AppCache.php"
        ]
    },
    "autoload-dev": {
        "psr-4": {
            "Tests\\": "tests/"
        }
    },
    "require": {
        "php": "^7.1",
        "ext-redis": "*",
        "symfony/symfony": "3.2.*",
        "doctrine/orm": "^2.5",
        "doctrine/doctrine-bundle": "^1.6",
        "doctrine/doctrine-cache-bundle": "^1.2",
        "doctrine/doctrine-migrations-bundle": "^1.2",
        "stof/doctrine-extensions-bundle": "^1.2",
        "league/tactician-doctrine": "^1.0",
        "ramsey/uuid-doctrine": "^1.2",
        "symfony/swiftmailer-bundle": "^2.3",
        "symfony/monolog-bundle": "~3.1",
        "symfony/polyfill-apcu": "^1.0",
        "sensio/distribution-bundle": "^5.0",
        "sensio/framework-extra-bundle": "^3.0.2",
        "incenteev/composer-parameter-handler": "^2.0",
        "league/tactician-bundle": "^0.4",
        "friendsofsymfony/rest-bundle": "~2.2",
        "nelmio/api-doc-bundle": "^2.11",
        "jms/serializer-bundle": "^1.1",
        "jms/di-extra-bundle": "^1.8",
        "guzzlehttp/guzzle": "6.2.2",
        "mopa/bootstrap-bundle": "~3.0",
        "symfony/assetic-bundle": "^2.8",
        "friendsofsymfony/jsrouting-bundle": "^1.6",
        "bmatzner/fontawesome-bundle": "~4.7",
        "tdn/php-types": "dev-develop as 3.0.0",
        "kzykhys/git": "v0.1.2",
        "cache/apcu-adapter": "~0.2",
        "cache/redis-adapter": "~0.4",
        "cache/cache-bundle": "~0.5",
        "cache/adapter-bundle": "~0.4",
        "cache/psr-6-doctrine-bridge": "^3.0",
        "kachkaev/assets-version-bundle": "~2.0",
        "e-moe/guzzle6-bundle": "~1.1",
        "composer/semver": "^1.4",
        "typo3/class-alias-loader": "dev-master",
        "oomphinc/composer-installers-extender": "^1.1",
        "bower-asset/bootstrap": "~3.3.0",
        "bower-asset/jquery": "~3.2",
        "bower-asset/jqueryui": "~1.12",
        "bower-asset/moment": "~2.18",
        "bower-asset/moment-timezone": "~0.5",
        "bower-asset/datatables": "~1.10",
        "bower-asset/remarkable-bootstrap-notify": "~3.1",
        "bower-asset/animate.css": "~3.1",
        "bower-asset/hint.css": "~2.5",
        "bower-asset/SpinKit": "~1.2",
        "bower-asset/bootstrap3-dialog": "~1.35",
        "bower-asset/async": "~2.4",
        "npm-asset/colors.css": "~3.0"
    },
    "require-dev": {
        "phploc/phploc": "^3.0",
        "sensio/generator-bundle": "^3.1",
        "dephpend/dephpend": "dev-develop",
        "henrikbjorn/lurker": "@stable",
        "doctrine/doctrine-fixtures-bundle": "^2.3",
        "symfony/phpunit-bridge": "dev-master",
        "friendsofphp/php-cs-fixer": "~2.0",
        "squizlabs/php_codesniffer": "~2.7",
        "phpunit/phpunit": "~6.1",
        "sebastian/phpcpd": "~3.0",
        "mockery/mockery": "~0.9",
        "liip/functional-test-bundle": "~1.7",
        "jakub-onderka/php-parallel-lint": "0.*",
         "phpmd/phpmd" : "@stable",
        "consolidation/robo": "~1.0",
        "seld/jsonlint": "~1.6",
        "kevinlebrun/colors.php": "0.*",
        "rskuipers/php-assumptions": "dev-master",
        "povils/phpmnd": "~1.0",
        "escapestudios/symfony2-coding-standard": "^2.10",
        "tm/tooly-composer-script": "^1.2"
    },
    "scripts": {
        "symfony-scripts": [
            "Incenteev\\ParameterHandler\\ScriptHandler::buildParameters",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget",
            "Tooly\\ScriptHandler::installPharTools"
        ],
        "post-install-cmd": [
            "@symfony-scripts"
        ],
        "post-update-cmd": [
            "@symfony-scripts"
        ]
    },
    "config": {
        "process-timeout": 10000,
        "preferred-install": "dist",
        "github-protocols": ["https"],
        "bin-dir": "bin",
        "platform": {
            "php": "7.1"
        },
        "fxp-asset": {
            "installer-paths": {
                "vcs-driver-options": {
                    "github-no-api": true
                },
                "asset-pattern-skip-version": "(-build|-patch)",
                "npm-asset-library": "web/vendor",
                "bower-asset-library": "web/vendor"
            }
        }
    },
    "extra": {
        "installer-types": ["bower-asset", "npm-asset"],
        "installer-paths": {
            "web/vendor/{$vendor}/{$name}/": [
                "type:bower-asset",
                "type:npm-asset"
            ]
        },
        "symfony-app-dir": "app",
        "symfony-bin-dir": "bin",
        "symfony-var-dir": "var",
        "symfony-web-dir": "web",
        "symfony-tests-dir": "tests",
        "symfony-assets-install": "relative",
        "incenteev-parameters": {
            "file": "app/config/parameters.yml"
        },
        "tools": {
            "phpmd-extension": {
                "url": "https://github.com/mi-schi/phpmd-extension/releases/download/stable/phpmd-extension.phar",
                "only-dev": true,
                "force-replace": true
            },
            "phpmetrics": {
                "url": "https://github.com/phpmetrics/PhpMetrics/releases/download/v2.2.0/phpmetrics.phar",
                "only-dev": true,
                "force-replace": true
            },
            "chag": {
                "url": "https://raw.githubusercontent.com/vpassapera/chag/master/chag",
                "only-dev": true,
                "force-replace": true
            }
        },
        "typo3/class-alias-loader": {
            "class-alias-maps": [
                "src/AppBundle/AliasMap.php"
            ],
            "always-add-alias-loader": true
        }
    },
    "repositories": [
        {
            "type": "composer",
            "url": "https://asset-packagist.org"
        }
    ]
}
```

**Comment by alexislefebvre** [comment]
@HeahDude Can we update the Travis CI test suite to Ubuntu Trusty so that you'll have to rebase your PR after the update?

**Comment by HeahDude** [comment]
Sure, no problem :).

**Comment by cacahouete78** [comment]
Merge soon ?

**Comment by alexislefebvre** [comment]
@cacahouete78 See the previous comment: https://github.com/liip/LiipFunctionalTestBundle/pull/322#issuecomment-305283622

PHPUnit 6.1 is not installed on Travis CI, so we don't know if these changes work.

**Comment by Aerendir** [comment]
So, which is the solution to check if the pr works?

**Comment by Aerendir** [comment]
@vpassapera Are you using the `typo3/class-alias-loader` just to run phpunit 6.1?

**Comment by Aerendir** [comment]
Anyway, the current PHPUnit version is 6.2...

**Comment by Aerendir** [comment]
I confirm that the workaround proposed by @vpassapera works well.

Just add this to `composer.json`:

```json
{
    ...
    "require-dev": {
        ...
        "typo3/class-alias-loader": "dev-master"
    },
    ...
    "extra": {
        ...
        "typo3/class-alias-loader": {
            "class-alias-maps": [
                "tests/AliasMap.php"
            ],
            "always-add-alias-loader": true
        }
    },
    ...
}
```

The alias map is the one proposed by @vpassapera and I put it into `tests` folder (Symfony 3.3):

```php
<?php
// tests/AliasMap.php

return [
    'PHPUnit_Framework_TestCase' => PHPUnit\Framework\TestCase::class,
];

```

**Comment by alexislefebvre** [comment]
@Aerendir

> So, which is the solution to check if the pr works?

I tried to use Ubuntu Trusty on Travis CI, but then I encountered an error with MySQL: https://github.com/alexislefebvre/LiipFunctionalTestBundle/pull/49

The MySQL error is a known problem: https://github.com/liip/LiipFunctionalTestBundle/pull/327#issuecomment-305974106

So I'm stuck ah the moment, feel free to offer another solution.

**Comment by Aerendir** [comment]
Unfortunately I haven't alternatives...

The only way at the moment, to make our apps running, is to use the workaround proposed by @vpassapera that works very well... also if it is a workaround...

We have to be patient...

**Comment by z38** [comment]
I just [tested](https://travis-ci.org/z38/LiipFunctionalTestBundle/builds/243711048) the last commit fcb1f18 again: PHPUnit 6.2 is now installed correctly on precise. While this did uncover some failing tests (confirmed locally), this opens an other option instead of waiting for #327: Ignore HHVM errors until MySQL is usable again for precise.

**Comment by alexislefebvre** [comment]
@z38 Thanks for the test, that's nice. Did updating the `composer.json` was enough to install PHPUnit 6.2? Or is it related to a change on Travis CI?

**Comment by alexislefebvre** [comment]
@HeahDude Following the suggestion from @z38, I ran the tests again. It fails now that PHPUnit 6.2 is installed. About HHVM, I think that we can remove it from the test matrix.

Now we have to find a solution so that tests run with PHPUnit 6 too…

**Comment by alexislefebvre** [comment]
I'm testing aliasing old classes to new classes: https://github.com/alexislefebvre/LiipFunctionalTestBundle/pull/50

**Comment by vpassapera** [comment]
@Aerendir Glad the solution I posted worked out for ya. Although to be fair, I didn't come up with it, pretty sure I found that floating on the internet somewhere (can't remember where though).

Aliasing old classes to new ones should work for the most part. Haven't tried going in depth with every single edge case though.

**Comment by alexislefebvre** [comment]
@HeahDude I added a commit that fix tests and I rebased this branch. It it OK for you?

**Comment by HeahDude** [comment]
@alexislefebvre Looks great! Thank you for taking care of it :), I'm sorry I've been very busy last days without internet most of the time.

**Comment by alexislefebvre** [comment]
@HeahDude no problem, thanks to you and everyone who contributed. I'll publish a new release today.
